src/pytorch/pme/pme.py

```diff
@@ -0,0 +1,194 @@
+        for i in range(1, order-1):
+            data[order-i-1] = (i*data[order-i-2]+(order-i)*data[order-i-1])/(order-1)
+        data[0] /= order-1
+        bsplines_data[1:order+1] = data
+
+        # Evaluate the actual bspline moduli for X/Y/Z.
+
+        self.moduli = []
+        for ndata in (gridx, gridy, gridz):
+            m = torch.zeros(ndata, dtype=torch.float32)
+            for i in range(ndata):
+                arg = (2*torch.pi*i/ndata)*torch.arange(ndata)
+                sc = torch.sum(bsplines_data[:ndata]*torch.cos(arg))
+                ss = torch.sum(bsplines_data[:ndata]*torch.sin(arg))
+                m[i] = sc*sc + ss*ss
+            for i in range(ndata):
+                if m[i] < 1e-7:
+                    m[i] = (m[(i-1+ndata)%ndata]+m[(i+1)%ndata])*0.5
+            self.moduli.append(m)
+
+    def compute_direct(self, positions: torch.Tensor, charges: torch.Tensor, cutoff: float, box_vectors: torch.Tensor, max_num_pairs: int = -1):
+        """Compute the direct space energy.
+
+        Parameters
+        ----------
+        positions: torch.Tensor
+            a 2D tensor of shape `(atoms, 3)` containing the positions of the atoms
+        charges: torch.Tensor
+            a 1D tensor of length `atoms` containing the charge of each atom
+        cutoff: float
+            the cutoff distance to use when computing the direct space term
+        box_vectors: torch.Tensor
+            The vectors defining the periodic box.  This must have shape `(3, 3)`,
+            where `box_vectors[0] = a`, `box_vectors[1] = b`, and `box_vectors[2] = c`.
+        max_num_pairs: int, optional
+            Maximum number of pairs (total number of neighbors). If set to `-1` (default),
+            all possible combinations of atom pairs are included.
+
+        Returns
+        -------
+        the energy of the direct space term
+        """
+        if positions.dim() != 2 or positions.shape[1] != 3:
+            raise ValueError('positions must have shape (atoms, 3)')
+        if charges.dim() != 1:
+            raise ValueError('charges must be 1D')
+        if positions.shape[0] != self.exclusions.shape[0] or charges.shape[0] != self.exclusions.shape[0]:
+            raise ValueError('positions, charges, and exclusions must all have the same length')
+        if box_vectors.dim() != 2 or box_vectors.shape[0] != 3 or box_vectors.shape[1] != 3:
+            raise ValueError('box_vectors must have shape (3, 3)')
+        if (cutoff <= 0):
+            raise ValueError('cutoff must be positive')
+        neighbors, deltas, distances, number_found_pairs = getNeighborPairs(positions, cutoff, max_num_pairs, box_vectors)
+        self.exclusions = self.exclusions.to(positions.device)
+        return torch.ops.pme.pme_direct(positions, charges, neighbors, deltas, distances, self.exclusions, self.alpha, self.coulomb)
+
+    def compute_reciprocal(self, positions: torch.Tensor, charges: torch.Tensor, box_vectors: torch.Tensor):
```

Comment:
Besides testing purposes, is there any situation in which one would call compute_direct() and compute_reciprocal() for any other thing apart from subtracting each other? 
I believe it would be useful for the PmeModule you wrote for the test to be defined in this file instead.

Reply:
One common situation is multiple time step integrators, where you evaluate the direct space part more often than the reciprocal space part.  I also want to experiment with models where I include only the reciprocal space part, which covers long range interactions for which point charges are a pretty good approximation, while letting the neural network try to learn a more accurate short range part.